src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1142,31 +1151,63 @@ Node* VectorNode::degenerate_vector_rotate(Node* src, Node* cnt, bool is_rotate_
   // later swap them in case of left rotation.
   Node* shiftRCnt = NULL;
   Node* shiftLCnt = NULL;
-  if (cnt->is_Con() && cnt->bottom_type()->isa_int()) {
-    // Constant shift case.
-    int shift = cnt->get_int() & shift_mask;
+  const TypeInt* cnt_type = cnt->bottom_type()->isa_int();
+  bool is_binary_vector_op = false;
+  if (cnt_type && cnt_type->is_con()) {
+    // Constant shift.
+    int shift = cnt_type->get_con() & shift_mask;
     shiftRCnt = phase->intcon(shift);
     shiftLCnt = phase->intcon(shift_mask + 1 - shift);
-  } else {
-    // Variable shift case.
+  } else if (VectorNode::is_invariant_vector(cnt)) {
+    // Scalar variable shift, handle replicates generated by auto vectorizer.
     assert(VectorNode::is_invariant_vector(cnt), "Broadcast expected");
     cnt = cnt->in(1);
     if (bt == T_LONG) {
       // Shift count vector for Rotate vector has long elements too.
-      assert(cnt->Opcode() == Op_ConvI2L, "ConvI2L expected");
-      cnt = cnt->in(1);
+      if (cnt->Opcode() == Op_ConvI2L) {
+         cnt = cnt->in(1);
+      } else {
+         assert(cnt->bottom_type()->isa_long() &&
+                cnt->bottom_type()->is_long()->is_con(), "Long constant expected");
```

Comment:
Is this case from Vector API? I think in Java we can have only Int type for shifts.

Reply:
> Is this case from Vector API? I think in Java we can have only Int type for shifts.
This is to handle long rotates, SLP will insert a ConvI2L before broadcasting shift value into a long vector. In case of constant int value, ConvI2L::Value will create a constant of TypeLong and which is what is being handled here.

Reply:
Got it.